File: python/src/azoth/eos/reference/pr_mass_density.py
"""``eos.pr_mass_density`` - mass density from a molar volume.

```text
rho = M/v
```

Spec: ``specs/calcs/eos/pr_mass_density.toml``, which carries why one division earns
a calc and the ``g/mol``-versus-``kg/mol`` trap that the units layer cannot catch.
"""

from __future__ import annotations

from azoth._registry_gen import spec as _spec_for
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import PrMassDensityResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning

CALC_ID = "eos.pr_mass_density"


def pr_mass_density(M: Q, v: Q) -> PrMassDensityResult:
    """Mass density from a molar mass and a molar volume.

    Args:
        M: molar mass. **Supplied by the caller** - this library ships no component
            data. Note the unit: ``kg/mol``, not the ``g/mol`` tables usually quote.
        v: molar volume, from :func:`azoth.eos.pr_molar_volume`.

    Returns:
        The mass density, in ``kg/m**3``.

    Raises:
        OutOfRangeError: if ``M <= 0`` or ``v <= 0``. Both are strictly positive for
            any real substance and state, and ``v`` is a divisor.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> r = pr_mass_density(q(0.0440956, "kg/mol"), q(0.0018317107825229842, "m**3/mol"))
        >>> round(r.rho.magnitude, 15)
        24.073451125981286
    """
    spec = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    values = {"M": input_to_si(spec, "M", M), "v": input_to_si(spec, "v", v)}

    apply_checks(checks.on_input, values.get, warnings)

    rho = values["M"] / values["v"]

    apply_checks(checks.derived, lambda name: rho if name == "rho" else None, warnings)

    return PrMassDensityResult(rho=from_si(rho, "kg/m**3"), warnings=tuple(warnings))

File: crates/azoth-eos/src/pr_mass_density.rs
//! `eos.pr_mass_density` - mass density from a molar volume.
//!
//! ```text
//! rho = M/v
//! ```
//!
//! Spec: `specs/calcs/eos/pr_mass_density.toml`, which carries why one division earns
//! a calc and the `g/mol`-versus-`kg/mol` trap that the units layer cannot catch.

use azoth_core::units::{MolarMass, MolarVolume, kilograms_per_cubic_meter};
use azoth_core::{Result, apply_checks};

use crate::results::PrMassDensityResult;
use crate::spec_gen;

/// Mass density from a molar mass and a molar volume.
///
/// `M` is the caller's - this library ships no component data - and `v` comes from
/// [`crate::pr_molar_volume`].
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `M <= 0` or `v <= 0`. Both are
///   strictly positive for any real substance and state, and `v` is a divisor.
///
/// # Example
/// ```
/// use azoth_core::units::{cubic_meters_per_mole, kilograms_per_mole};
/// use azoth_eos::pr_mass_density;
///
/// let r = pr_mass_density(
///     kilograms_per_mole(0.0440956),
///     cubic_meters_per_mole(0.0018317107825229842),
/// )?;
/// assert!((r.rho.value - 24.073451125981286).abs() < 1e-12);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
#[allow(non_snake_case)] // `M` is the symbol in the equation
pub fn pr_mass_density(M: MolarMass, v: MolarVolume) -> Result<PrMassDensityResult> {
    let spec = &spec_gen::PR_MASS_DENSITY_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "M" => Some(M.value),
            "v" => Some(v.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let rho = M.value / v.value;

    apply_checks(
        spec.derived_checks(),
        |quantity| (quantity == "rho").then_some(rho),
        &mut warnings,
    )?;

    Ok(PrMassDensityResult {
        rho: kilograms_per_cubic_meter(rho),
        warnings,
    })
}
